drivers/serial/uart_nrfx_uart.c

```diff
@@ -291,6 +318,344 @@ static int uart_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+		nrf_uart_rxd_get(uart0_addr);
+	} else {
+		if (uart0_cb.rx_counter == 0) {
+			event.type = UART_RX_BUF_REQUEST;
+			user_callback(&event);
+		}
+		uart0_cb.rx_buffer[uart0_cb.rx_counter] =
+			nrf_uart_rxd_get(uart0_addr);
+		uart0_cb.rx_counter++;
+		if (uart0_cb.rx_timeout == K_NO_WAIT) {
+			rx_rdy_evt();
+		} else if (uart0_cb.rx_timeout != K_FOREVER) {
+			k_delayed_work_submit(&uart0_cb.rx_timeout_work,
+					      uart0_cb.rx_timeout);
+		}
+	}
+
+	if (uart0_cb.rx_buffer_length == uart0_cb.rx_counter) {
+		k_delayed_work_cancel(&uart0_cb.rx_timeout_work);
+		rx_rdy_evt();
+
+		if (uart0_cb.rx_secondary_buffer_length) {
+			buf_released_evt();
+			/* Switch to secondary buffer. */
+			uart0_cb.rx_buffer_length =
+				uart0_cb.rx_secondary_buffer_length;
+			uart0_cb.rx_buffer = uart0_cb.rx_secondary_buffer;
+			uart0_cb.rx_secondary_buffer_length = 0;
+			uart0_cb.rx_counter = 0;
+			uart0_cb.rx_offset = 0;
+
+			event.type = UART_RX_BUF_REQUEST;
+			user_callback(&event);
+		} else {
+			uart_nrfx_rx_disable(dev);
+		}
+	}
+}
+
+static void tx_isr(void)
+{
+	uart0_cb.tx_counter++;
+	if (uart0_cb.tx_counter < uart0_cb.tx_buffer_length &&
+	    !uart0_cb.tx_abort) {
+#if	defined(DT_NORDIC_NRF_UART_UART_0_RTS_PIN) && \
+		defined(DT_NORDIC_NRF_UART_UART_0_CTS_PIN)
+		k_delayed_work_submit(&uart0_cb.tx_timeout_work,
+				uart0_cb.tx_timeout);
+#endif
+		nrf_uart_event_clear(uart0_addr, NRF_UART_EVENT_TXDRDY);
+
+		u8_t txd = uart0_cb.tx_buffer[uart0_cb.tx_counter];
+
+		nrf_uart_txd_set(uart0_addr, txd);
+	} else {
+#if	defined(DT_NORDIC_NRF_UART_UART_0_RTS_PIN) && \
+		defined(DT_NORDIC_NRF_UART_UART_0_CTS_PIN)
+		k_delayed_work_cancel(&uart0_cb.tx_timeout_work);
+#endif
+		nrf_uart_event_clear(uart0_addr, NRF_UART_EVENT_TXDRDY);
+		uart0_cb.tx_buffer_length = 0;
+		uart0_cb.tx_counter = 0;
+		struct uart_event event = {
+			.type = UART_TX_DONE,
+			.data.tx.buf = uart0_cb.tx_buffer,
+			.data.tx.len = uart0_cb.tx_counter
+		};
+		user_callback(&event);
+	}
+}
+
+#define UART_ERROR_FROM_MASK(mask) \
+	(mask & NRF_UART_ERROR_OVERRUN_MASK ? UART_ERROR_OVERRUN	\
+	 : mask & NRF_UART_ERROR_PARITY_MASK ? UART_ERROR_PARITY	\
+	 : mask & NRF_UART_ERROR_FRAMING_MASK ? UART_ERROR_FRAMING	\
+	 : mask & NRF_UART_ERROR_BREAK_MASK ? UART_BREAK		\
+	 : 0)
+
+static void error_isr(struct device *dev)
+{
+	k_delayed_work_cancel(&uart0_cb.rx_timeout_work);
+	nrf_uart_event_clear(uart0_addr, NRF_UART_EVENT_ERROR);
+
+	if (!uart0_cb.rx_enabled) {
+		nrf_uart_task_trigger(uart0_addr, NRF_UART_TASK_STOPRX);
+	}
+	struct uart_event event = {
+		.type = UART_RX_STOPPED,
+		.data.rx_stop.reason =
+			UART_ERROR_FROM_MASK(
+				nrf_uart_errorsrc_get_and_clear(uart0_addr)),
+		.data.rx_stop.data.len = uart0_cb.rx_counter
+					 - uart0_cb.rx_offset,
+		.data.rx_stop.data.offset = uart0_cb.rx_offset,
+		.data.rx_stop.data.buf = uart0_cb.rx_buffer
+	};
+
+	user_callback(&event);
+	/* Abort transfer. */
+	uart_nrfx_rx_disable(dev);
+}
+
+static void rxto_isr(void)
```

Comment:
In my list, "rxto" means "rx timeout". If nRF has hardware RX timeout, why do you use Zephyr work item above? All in all, I'd say this function needs comments, it's unclear what it does and how to someone who doesn't know specific details of the nRF chip(s) (while as the only driver so far, it'll serve as a reference for other MCUs).

Reply:
Added a comment regarding that